pkg/endpoint/bpf.go

```diff
@@ -1202,23 +1202,31 @@ func (e *Endpoint) applyPolicyMapChanges() (proxyChanges bool, err error) {
 		for _, visMeta := range e.visibilityPolicy.Ingress {
 			proxyID := policy.ProxyID(e.ID, visMeta.Ingress, visMeta.Proto.String(), visMeta.Port)
 			if redirectPort, exists := e.realizedRedirects[proxyID]; exists && redirectPort != 0 {
-				e.desiredPolicy.PolicyMapState.AddVisibilityKeys(e, redirectPort, visMeta, adds, deletes)
+				e.desiredPolicy.PolicyMapState.AddVisibilityKeys(e, redirectPort, visMeta, adds, nil)
 			}
 		}
 		for _, visMeta := range e.visibilityPolicy.Egress {
 			proxyID := policy.ProxyID(e.ID, visMeta.Ingress, visMeta.Proto.String(), visMeta.Port)
 			if redirectPort, exists := e.realizedRedirects[proxyID]; exists && redirectPort != 0 {
-				e.desiredPolicy.PolicyMapState.AddVisibilityKeys(e, redirectPort, visMeta, adds, deletes)
+				e.desiredPolicy.PolicyMapState.AddVisibilityKeys(e, redirectPort, visMeta, adds, nil)
 			}
 		}
 	}
 
 	// Add policy map entries before deleting to avoid transient drops
-	for keyToAdd, entry := range adds {
+	for keyToAdd := range adds {
 		// AddVisibilityKeys() records changed keys in both 'deletes' (old value) and 'adds' (new value).
 		// Remove the key from 'deletes' to keep the new entry.
 		delete(deletes, keyToAdd)
 
+		entry, exists := e.desiredPolicy.PolicyMapState[keyToAdd]
+		if !exists {
+			e.getLogger().WithFields(logrus.Fields{
+				logfields.AddedPolicyID: keyToAdd,
+			}).Warn("Tried adding policy map key not in policy")
+			continue
```

Comment:
Is this an expected condition, reliant on a race condition? Or is it an unexpected condition that would demonstrate a bug? I'm mostly wondering whether there's any possibility that, for instance, in a high churn cluster with identities being introduced & removed all the time, could this warning message be triggered regularly? Could it flood the logs? Is the plan for us to gather any instances of this warning and use them to investigate further issues?

Reply:
The main motivation of changing `MapState` type to `Keys` is to avoid hitting this condition as we don't allow callers to erroneously mutate owners of a `Key`. `consumeMapChanges` (ideally) should be the only one doing book-keeping of owners and hence removal of keys from the policy map.
So this condition hopefully allows us to identify the cases where this doesn't hold true during the add operation. 
@jrajahalme Thinking out loud - do we need a reciprocal condition in the delete path? A key is being deleted but it exists in the policy map?

Reply:
This should be an internal (to the agent) error condition. I'll add unit tests covering addition and deletion of an entry on the same round of incremental updates. This could happen if a global identity is created and then deleted before the endpoints had a change to regenerate.